mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala

```diff
@@ -237,13 +247,33 @@ class PrefixSpan private (
    */
   private def getPatternsInLocal(
       minCount: Long,
-      data: RDD[(List[Int], Iterable[Array[Int]])]): RDD[(List[Int], Long)] = {
+      data: RDD[(List[Set[Int]], Iterable[List[Set[Int]]])]): RDD[(List[Set[Int]], Long)] = {
     data.flatMap {
-      case (prefix, projDB) =>
-        LocalPrefixSpan.run(minCount, maxPatternLength, prefix.toList.reverse, projDB)
-          .map { case (pattern: List[Int], count: Long) =>
-          (pattern.reverse, count)
-        }
+      case (prefix, projDB) => LocalPrefixSpan.run(minCount, maxPatternLength, prefix, projDB)
+    }
+  }
+
+}
+
+object PrefixSpan {
+  private[fpm] val DELIMITER = -1
+
+  private[fpm] def splitAtDelimiter(pattern: List[Int]): List[Set[Int]] = {
+    pattern.span(_ != DELIMITER) match {
+      case (x, xs) if xs.length > 1 => x.toSet :: splitAtDelimiter(xs.tail)
+      case (x, xs) => List(x.toSet)
     }
   }
+
+  private[fpm] def insertDelimiters(sequence: List[Set[Int]]): List[Int] = {
+    // TODO: avoid allocating new arrays when appending
+    sequence.zip(Seq.fill(sequence.size)(PrefixSpan.DELIMITER))
```

Comment:
Use `ArrayBuilder` instead of functional style. Creating a sequence of delimiters is not efficient. As a reference, Scala also use builder in `mkString`: https://github.com/scala/scala/blob/2.11.x/src/library/scala/collection/TraversableOnce.scala#L351.